src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicTlsSecret.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Quic;
+using static Microsoft.Quic.MsQuic;
+
+namespace System.Net.Quic;
+
+internal sealed class MsQuicTlsSecret : IDisposable
+{
+    private static readonly int ClientRandomOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientRandom");
+    private static readonly int ClientEarlyTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientEarlyTrafficSecret");
+    private static readonly int ClientHandshakeTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientHandshakeTrafficSecret");
+    private static readonly int ServerHandshakeTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ServerHandshakeTrafficSecret");
+    private static readonly int ClientTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientTrafficSecret0");
+    private static readonly int ServerTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ServerTrafficSecret0");
+
+    public QUIC_TLS_SECRETS tlsSecret;
+    private GCHandle gcHandle;
+
+    public unsafe MsQuicTlsSecret(MsQuicContextSafeHandle handle)
+    {
+        tlsSecret = default;
+        gcHandle = GCHandle.Alloc(this, GCHandleType.Pinned);
+
+        fixed (void* ptr = &tlsSecret)
+        {
+            // Best effort. Ignores failures.
+            MsQuicApi.Api.SetParam(handle, QUIC_PARAM_CONN_TLS_SECRETS, (uint)sizeof(QUIC_TLS_SECRETS), ptr);
+        }
+    }
+
+    public void WriteSecret(string fileName)
```

Comment:
Why the split into 2 public methods?

Reply:
The screen one will open file for each Quic connection and allocate. And I did not want to drag the stream to `QuicConnection` right now and I did not want to worry about flushing in log running process. So I split them even if there is currently no benefit. I can merge them together but this was my motivation for keeping them separate.